src/book.py

```diff
@@ -180,39 +217,79 @@ def tokenize(self, on, rem_stopwords=True, stopword_lang='english',
-            tokens = sent_tokenize(self._text)
+            tokens = self.tokenize('sent', rem_stopwords)
         else:
             raise KeyError(
-                "Arugument 'on' must refer to character, word, or sentence")
-
-        if non_cont:
-            # TODO add functionality for non-continuous random sampling
-            pass
+                "Argument 'on' must refer to character, word, or sentence")
+            
+        assert groups*length <= len(tokens)
+        return_array = []
+
+        if randomized:
+            start = random.randint(0, len(tokens) - groups*length)
+            tokens = tokens[start:]
+            start = 0
         else:
-            if 'char' in on.lower():
-                start = random.randint(0, len(self._text) - length)
-                snippet = self._text[start:start + length]
-            elif 'word' in on.lower():
-                start = random.randint(0, len(tokens) - length)
-                snippet = (" ").join(tokens[start:start + length])
-            elif 'sent' in on.lower():
-                start = random.randint(0, len(tokens) - length)
-                snippet = (" ").join(tokens[start:start + length])
+            start = 0
+        
+        if non_cont == False:
+            for gr in range(groups):
+                snippet = BookText(rawtext=''.join((" ").join(tokens[length*gr:length*(gr+1)])), 
+                               author = self.author, title = self.title, meta = self.meta)
+                return_array.append(snippet)
+        else:
+            for gr in range(groups): 
+                snippet = BookText(rawtext=''.join((" ").join(tokens[start:start+length])), 
+                               author = self.author, title = self.title, meta = self.meta)
+                snippet._text.rstrip()    
+                return_array.append(snippet)
+                
+                for index in sorted(range(start, start+length), reverse=True):
+                    if (len(tokens)-length > 0):
+                        del tokens[index] 
+                if (random_seed is not None):
+                    random.seed(random_seed + gr + 1)
+            
+                if (len(tokens)-length > 0):
+                    start = random.randint(1, len(tokens)-length)
+                else:
+                    start = 0
+        
+        dummy_string = str()
+        for book in return_array:
+            dummy_string += book._text + str(' ')
+            
+        dummy_string.strip()
+
+        snip = BookText(rawtext=dummy_string, author=self.author, title=self.title, meta=self.meta)
+        
         if inplace:
-            self._text = snippet
+            self._text = snip._text
         else:
-            return BookText(rawtext=snippet, author=self.author,
-                            title=self.title, meta=self.meta)
+            if ret_as_arr == False:
+                return snip
+            else:
+                return return_array
```

Comment:
Strictly speaking `return_array` isn't an array at least in the way I think of it. It is a list object. You could either use `np.asarray(return_array)` here or instantiate `return_array` as an empty array of length `groups`.

Reply:
return_array was instantiated in the body.

Reply:
Yes but it is instantiated as `return_array = []` which is not a numpy array. This won't break anything, but it goes against expectations when the argument `ret_as_arr=True` gives you a list back.

Reply:
I see. Ok.

Reply:
In the latest commit, it is returned as a numpy array.